### PR TITLE
server: add support for sorting zones in UI/API

### PR DESCRIPTION
**Pull request by rhtyd**
*Problem*: Not able to configure a sort order for the zones that are listed in various views in the UI.

*Root Cause:* There is no mechanism to accept sort key for existing zones or UI widget, that would allow to listing zones in the UI in a certain order.

*Solution*: The order of zones in listed in various views in the UI can now be configured through the newly added “sort_key” field added for the zone. It can be set using updateZone API by providing “sort_key” parameter for a zone, or by reordering the items in the zones list in the UI. UI has been updated to show ordering controls in zones list view. Database changes include updating table “data_center” by adding “sort_key” column (containing integer values and defaults to zero).

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

51236319-da9d2a80-1997-11e9-9c23-b8a4e5e27ccd

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: api/src/main/java/org/apache/cloudstack/api/command/admin/zone/UpdateZoneCmd.java [hunk 1]
```diff
@@ -95,6 +95,9 @@
     @Parameter(name = ApiConstants.LOCAL_STORAGE_ENABLED, type = CommandType.BOOLEAN, description = "true if local storage offering enabled, false otherwise")
     private Boolean localStorageEnabled;
 
+    @Parameter(name = ApiConstants.SORT_KEY, type = CommandType.INTEGER, description = "sort key of the disk offering, integer")
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/zone/UpdateZoneCmd.java]: this is not the sortkey for the diskoffering, is it?

File: api/src/main/java/org/apache/cloudstack/query/QueryService.java [hunk 2]
```diff
@@ -85,9 +85,12 @@
 public interface QueryService {
 
     // Config keys
-    static final ConfigKey<Boolean> AllowUserViewDestroyedVM = new ConfigKey<Boolean>("Advanced", Boolean.class, "allow.user.view.destroyed.vm", "false",
+    ConfigKey<Boolean> AllowUserViewDestroyedVM = new ConfigKey<>("Advanced", Boolean.class, "allow.user.view.destroyed.vm", "false",
             "Determines whether users can view their destroyed or expunging vm ", true, ConfigKey.Scope.Account);
 
+    ConfigKey<Boolean> SortKeyAscending = new ConfigKey<>("Advanced", Boolean.class, "sortkey.algorithm", "true",
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/query/QueryService.java]: so true means that all lists for all users are sorted ascending? (if one of template, disk offering, service offering, network offering or zone)

File: api/src/main/java/org/apache/cloudstack/api/command/admin/zone/UpdateZoneCmd.java (hunk 1 above)
Comment (anuragaw) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/zone/UpdateZoneCmd.java]: Yes @DaanHoogland , I'll fix.

File: api/src/main/java/org/apache/cloudstack/query/QueryService.java (hunk 2 above)
Comment (anuragaw) [reply api/src/main/java/org/apache/cloudstack/query/QueryService.java]: Yes @DaanHoogland . That's how global config seems to have been designed and used.

Comment (rhtyd) [reply api/src/main/java/org/apache/cloudstack/query/QueryService.java]: Can you rectify the description to reflect the general use-case of the config key @anuragaw ?

Comment (anuragaw) [reply api/src/main/java/org/apache/cloudstack/query/QueryService.java]: Fixing @rhtyd

**Comment by rhtyd** [comment]
@anuragaw can you help fix the conflicts?

**Comment by rhtyd** [comment]
Thanks @shwstppr 
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Lgtm @anuragaw

**Pull request merged**